mass-deployer/pkg/mass-deployer/deployer.go

```diff
@@ -0,0 +1,281 @@
+package deployer
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"sync"
+	"time"
+
+	"gopkg.in/yaml.v3"
+
+	"github.com/rs/zerolog/log"
+	"github.com/threefoldtech/tfgrid-sdk-go/grid-client/deployer"
+	"github.com/threefoldtech/tfgrid-sdk-go/grid-client/workloads"
+	"github.com/threefoldtech/tfgrid-sdk-go/grid-proxy/pkg/types"
+	"github.com/threefoldtech/zos/pkg/gridtypes"
+)
+
+type groupDeploymentsInfo struct {
+	vmDeployments      []*workloads.Deployment
+	networkDeployments []*workloads.ZNet
+	deploymentsInfo    []vmDeploymentInfo
+}
+
+type vmDeploymentInfo struct {
+	nodeID         uint32
+	vmName         string
+	deploymentName string
+}
+
+func RunDeployer(cfg Config) error {
+	ctx := context.Background()
+
+	tfPluginClient, err := setup(cfg)
+	if err != nil {
+		return fmt.Errorf("failed to create deployer: %v", err)
+	}
+
+	groupsNodes, failed := filterNodes(tfPluginClient, cfg.NodeGroups, ctx)
+	failedGroups := failed
+	passedGroups := map[string][]string{}
+
+	groupsDeployments := parseVMs(tfPluginClient, cfg.Vms, groupsNodes, cfg.SSHKeys)
+
+	deploymentStart := time.Now()
+
+	for nodeGroup, deployments := range groupsDeployments {
+		info, err := massDeploy(tfPluginClient, ctx, deployments)
+		if err != nil {
+			failedGroups[nodeGroup] = err
+		} else {
+			passedGroups[nodeGroup] = info
+		}
+	}
+
+	log.Info().Msgf("deployment took %s", time.Since(deploymentStart))
+
+	if len(passedGroups) > 0 {
+		fmt.Println("ok:")
+	}
+	for group, info := range passedGroups {
+		fmt.Printf("%s: \n%+v\n", group, info)
+	}
+
+	if len(failedGroups) > 0 {
+		fmt.Println("error:")
```

Comment:
errors always go to stderr `fmt.Fprintln(os.Stderr, .....` and same for the failed groups